deep_qa/training/models.py

```diff
@@ -28,6 +35,62 @@ def _summary_with_mask_info(self):
         flattened_layers = getattr(self, 'flattened_layers', self.layers)
         print_summary_with_masking(flattened_layers, getattr(self, 'container_nodes', None))
 
+    @overrides
+    def compile(self, params: Params):  # pylint: disable=arguments-differ
+        # pylint: disable=attribute-defined-outside-init
+        """
+        The only reason we are overriding this method is because keras automatically wraps
+        our tensorflow optimiser in a keras wrapper, which we don't want. We override the
+        only method in ``Model`` which uses this attribute, ``_make_train_function``, which
+        raises an error if compile is not called first.
+
+        As we move towards using a Tensorflow first optimisation loop, more things will be
+        added here which add functionality to the way Keras runs tensorflow Session calls.
+
+        """
+        optimizer = params.get('optimizer')
+        self.gradient_clipping = params.pop("gradient_clipping", None)
+        print(self.gradient_clipping.as_dict())
+        super(DeepQaModel, self).compile(**params.as_dict())
+        self.optimizer = optimizer
+
+    @overrides
+    def _make_train_function(self):
+        # pylint: disable=attribute-defined-outside-init
+        """
+        We override this method so that we can use tensorflow optimisers directly.
+        This is desirable as tensorflow handles gradients of sparse tensors efficiently.
+        """
+        if not hasattr(self, 'train_function'):
+            raise RuntimeError('You must compile your model before using it.')
+        if self.train_function is None:
+            inputs = self._feed_inputs + self._feed_targets + self._feed_sample_weights
+            if self.uses_learning_phase and not isinstance(K.learning_phase(), int):
+                inputs += [K.learning_phase()]
+
+            # Here we override Keras to use tensorflow optimizers directly.
+            self.global_step = K.variable(0., name='global_step')
+            grads = tensorflow.gradients(self.total_loss, self._collected_trainable_weights)
+            if self.gradient_clipping is not None:
+                # Don't pop from the gradient clipping dict here as
+                # when we load models we need it to still be there.
+                clip_type = self.gradient_clipping.get("type")
+                clip_value = self.gradient_clipping.get("value")
+                if clip_type == 'clip_by_norm':
+                    grads, _ = tensorflow.clip_by_global_norm(grads, clip_value)
+                elif clip_type == 'clip_by_value':
+                    grads = [tensorflow.clip_by_value(x, -clip_value, clip_value) for x in grads]
+                else:
+                    raise ConfigurationError("{} is not a supported type of gradient clipping.".format(clip_type))
+
+            zipped_grads_with_weights = zip(grads, self._collected_trainable_weights)
+            # pylint: disable=no-member
```

Comment:
The problem is that model assigns `optimiser` in `compile`, but then we override it with a tensorflow optimiser. I didn't override compile, as it would require copying loads of code - this is probably something we could do eventually but for the moment this was considerably simpler, if a bit hacky. I'll add the enable after that line.